src/OmniSharp.Roslyn.CSharp/Services/Files/OnFilesChangedService.cs

```diff
@@ -1,32 +1,66 @@
 using System.Collections.Generic;
 using System.Composition;
+using System.IO;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
 using OmniSharp.FileWatching;
 using OmniSharp.Mef;
 using OmniSharp.Models;
 using OmniSharp.Models.FilesChanged;
 
+
 namespace OmniSharp.Roslyn.CSharp.Services.Files
 {
     [OmniSharpHandler(OmniSharpEndpoints.FilesChanged, LanguageNames.CSharp)]
     public class OnFilesChangedService : IRequestHandler<IEnumerable<Request>, FilesChangedResponse>
     {
         private readonly IFileSystemWatcher _watcher;
+        private OmniSharpWorkspace _workspace;
+        private object _lock = new object();
 
         [ImportingConstructor]
-        public OnFilesChangedService(IFileSystemWatcher watcher)
+        public OnFilesChangedService(IFileSystemWatcher watcher, OmniSharpWorkspace workspace)
         {
             _watcher = watcher;
+            _workspace = workspace;
         }
 
         public Task<FilesChangedResponse> Handle(IEnumerable<Request> requests)
         {
             foreach (var request in requests)
             {
+                RemoveRenamedFiles(request.FileName);
                 _watcher.TriggerChange(request.FileName);
             }
             return Task.FromResult(new FilesChangedResponse());
         }
+
+        private void RemoveRenamedFiles(string fileName)
+        {
+            lock (_lock)
+            {
+                var path = Path.GetDirectoryName(fileName);
+                foreach (var project in _workspace.CurrentSolution.Projects)
+                {
+                    if (!path.StartsWith(Path.GetDirectoryName(project.FilePath)))
+                    {
+                        continue;
+                    }
+
+                    foreach (var document in project.Documents)
+                    {
+                        if (!document.FilePath.Contains(path))
+                        {
+                            continue;
+                        }
+
+                        if(!File.Exists(document.FilePath) && _workspace.CurrentSolution.ContainsDocument(document.Id))
+                        {
+                            _workspace.TryApplyChanges(_workspace.CurrentSolution.RemoveDocument(document.Id));
```

Comment:
This is the wrong approach and can possible fail if multiple documents are removed from different projects. The reason for this is because `_worksapce.TryApplyChanges(...)` will update `_workspace.CurrentSolution`. However, the foreach loop will have received an enumerator from the original Solution.

I recommend the following approach:

1. Capture `_workspace.CurrentSolution` into a local variable.
2. Loop through all its projects and documents and make necessary changes to a new Solution.
3. After the loop, call `_workpsace.TryApplyChanges` with the new Solution if it is different than the old Solution.

Because all of this is done on immutable data structures (except the workspace itself), you shouldn't need the lock at all. Instead, it should be something like this (modulo errors because I'm tying into GitHub):

```C#
var directoryName = Path.GetDirectoryName(fileName);

var oldSolution = _workspace.CurrentSolution;
var newSolution = oldSolution;

foreach (var project in oldSolution.Projects)
{
    if (!directoryName.StartsWith(Path.GetDirectoryName(project.FilePath), StringComparison.OrdinalIgnoreCase))
    {
        continue;
    }

    foreach (var document in project.Documents)
    {
        if (fileName.Equals(document.FilePath, StringComparison.OrdinalIgnoreCase) &&
            !File.Exists(fileName))
        {
            newSolution = newSolution.RemoveDocument(document.Id);
        }
    }
}

if (oldSolution != newSolution)
{
   _workspace.TryApplyChanges(newSolution);
}
```